apis/r/R/SOMAExperimentAxisQuery.R

```diff
@@ -163,6 +168,527 @@ SOMAExperimentAxisQuery <- R6::R6Class(
+    #'
+    #' @param obsm_layer Name of array in \code{obsm} to load as the
+    #' cell embeddings
+    #' @param varm_layer Name of the array in \code{varm} to load as the  feature
+    #' loadings; will try to determine \code{varm_layer} from \code{obsm_layer}
+    #' @param obs_index Name of column in \code{obs} to add as cell names
+    #' @param var_index Name of column in \code{var} to add as feature names
+    #'
+    #' @return A \code{\link[SeuratObject]{DimReduc}} object
+    #'
+    to_seurat_reduction = function(
+      obsm_layer,
+      varm_layer = NULL,
+      obs_index = NULL,
+      var_index = NULL
+    ) {
+      .check_seurat_installed()
+      stopifnot(
+        "'obsm_layer' must be a single character value" = is_scalar_character(obsm_layer),
+        "'varm_layer' must be a single character value" = is.null(varm_layer) ||
+          is_scalar_character(varm_layer) ||
+          is_scalar_logical(varm_layer),
+        "one of 'obsm_layer' or 'varm_layer' must be provided" =
+          (is_scalar_character(obsm_layer) || is_scalar_logical(obsm_layer)) ||
+          (is_scalar_character(varm_layer) || is_scalar_logical(varm_layer)),
+        "'obs_index' must be a single character value" = is.null(obs_index) ||
+          (is_scalar_character(obs_index) && !is.na(obs_index)),
+        "'var_index' must be a single character value" = is.null(var_index) ||
+          (is_scalar_character(var_index) && !is.na(var_index))
+      )
+      # Check embeddings/loadings
+      ms_embed <- tryCatch(expr = self$ms$obsm$names(), error = null)
+      ms_load <- tryCatch(expr = self$ms$varm$names(), error = null)
+      if (is.null(ms_embed) && is.null(ms_load)) {
+        warning("No reductions present", call. = FALSE)
+        return(NULL)
+      }
+      if (is.null(ms_embed)) {
+        stop("No embeddings in obsm present", call. = FALSE)
+      }
+      names(ms_embed) <- .anndata_to_seurat_reduc(ms_embed)
+      if (is.null(ms_load) && !is.null(varm_layer)) {
+        msg <- "No loadings present in 'varm'"
+        if (is.null(obsm_layer)) {
+          stop(msg, call. = FALSE)
+        }
+        warning(msg, call. = FALSE, immediate. = TRUE)
+        varm_layer <- NULL
+      } else {
+        names(ms_load) <- .anndata_to_seurat_reduc(ms_load, 'loadings')
+      }
+      # Check provided names
+      if (!obsm_layer %in% c(ms_embed, names(ms_embed))) {
+        stop("Cannot find embeddings ", sQuote(obsm_layer), call. = FALSE)
+      }
+      if (is_scalar_character(varm_layer) && !varm_layer %in% c(ms_load, names(ms_load))) {
+        stop("Cannot find loadings in 'varm'", sQuote(varm_layer), call. = FALSE)
+      }
+      # Find Seurat name
+      seurat <- c(
+        .anndata_to_seurat_reduc(obsm_layer),
+        tryCatch(
+          expr = .anndata_to_seurat_reduc(varm_layer, 'loadings'),
+          error = null
+        )
+      )
+      if (length(seurat) == 2L && !identical(x = seurat[1L], y = seurat[2L])) {
```

Comment:
Might be slightly clearer as a named list or vector so downstream references can be expressed as `seurat$obsm_layer` (or similar) rather than `seurat[1L]`.